### PR TITLE
Kubernetes Response Engine, Part 7: Falcosidekick + Cloud Functions

### LINKED ISSUE #443: Creating Kubernetes Response Engine using GCE and Google Cloud Functions
/area blog

**What would you like to be added**:

Recently falcosidekick got a new output type called [Google Cloud Functions](https://github.com/falcosecurity/falcosidekick/pull/241). In this blog, we will try to demonstrate that how we can use this to implement a new type of Kubernetes Response Engine.

gcloudfalcov2 (1)

### PR DESCRIPTION
**Pull request by Dentrax**
Signed-off-by: Furkan <furkan.turkal@trendyol.com>
Co-authored-by: Batuhan <batuhan.apaydin@trendyol.com>

kre_gke_v3

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind user-interface

/kind content

> /kind translation

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area blog

> /area documentation

> /area videos

**What this PR does / why we need it**:

Please see the linked issue.

**Which issue(s) this PR fixes**:

Fixes #443

**Special notes for your reviewer**:
Currently W.I.P

### CODE REVIEW

**Comment by Dentrax** [comment]
Thank you so much for your support @developer-guy 💫

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+>
+> * [Kubernetes Response Engine, Part 1 : Falcosidekick + Kubeless]({{< ref "/blog/falcosidekick-reponse-engine-part-1-kubeless" >}})
+> * [Kubernetes Response Engine, Part 2 : Falcosidekick + OpenFaas]({{< ref "/blog/falcosidekick-reponse-engine-part-2-openfass" >}})
+> * [Kubernetes Response Engine, Part 3 : Falcosidekick + Knative]({{< ref "/blog/falcosidekick-reponse-engine-part-3-knative" >}})
+> * [Kubernetes Response Engine, Part 4 : Falcosidekick + Tekton]({{< ref "/blog/falcosidekick-reponse-engine-part-4-tekton" >}})
+> * [Kubernetes Response Engine, Part 5 : Falcosidekick + Argo]({{< ref "/blog/falcosidekick-reponse-engine-part-5-argo" >}})
+> * [Kubernetes Response Engine, Part 6 : Falcosidekick + Cloud Run]({{< ref "/blog/falcosidekick-reponse-engine-part-6-cloud-run" >}})
+---
+
+Earlier in this series we have seen how to use Argo, Tekton and Knative to trigger a resource after getting input from Falcosidekick.
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:20]: Earlier in this series, we have seen how to use Argo, Tekton, and Knative to trigger a resource after getting input from Falcosidekick.

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+> * [Kubernetes Response Engine, Part 1 : Falcosidekick + Kubeless]({{< ref "/blog/falcosidekick-reponse-engine-part-1-kubeless" >}})
+> * [Kubernetes Response Engine, Part 2 : Falcosidekick + OpenFaas]({{< ref "/blog/falcosidekick-reponse-engine-part-2-openfass" >}})
+> * [Kubernetes Response Engine, Part 3 : Falcosidekick + Knative]({{< ref "/blog/falcosidekick-reponse-engine-part-3-knative" >}})
+> * [Kubernetes Response Engine, Part 4 : Falcosidekick + Tekton]({{< ref "/blog/falcosidekick-reponse-engine-part-4-tekton" >}})
+> * [Kubernetes Response Engine, Part 5 : Falcosidekick + Argo]({{< ref "/blog/falcosidekick-reponse-engine-part-5-argo" >}})
+> * [Kubernetes Response Engine, Part 6 : Falcosidekick + Cloud Run]({{< ref "/blog/falcosidekick-reponse-engine-part-6-cloud-run" >}})
+---
+
+Earlier in this series we have seen how to use Argo, Tekton and Knative to trigger a resource after getting input from Falcosidekick. 
+Recently, Falcosidekick got a new output type support for [Cloud Functions](https://github.com/falcosecurity/falcosidekick/pull/241).
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:21]: Recently, Falcosidekick received a new output type support for [Cloud Functions](https://github.com/falcosecurity/falcosidekick/pull/241).

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+> * [Kubernetes Response Engine, Part 3 : Falcosidekick + Knative]({{< ref "/blog/falcosidekick-reponse-engine-part-3-knative" >}})
+> * [Kubernetes Response Engine, Part 4 : Falcosidekick + Tekton]({{< ref "/blog/falcosidekick-reponse-engine-part-4-tekton" >}})
+> * [Kubernetes Response Engine, Part 5 : Falcosidekick + Argo]({{< ref "/blog/falcosidekick-reponse-engine-part-5-argo" >}})
+> * [Kubernetes Response Engine, Part 6 : Falcosidekick + Cloud Run]({{< ref "/blog/falcosidekick-reponse-engine-part-6-cloud-run" >}})
+---
+
+Earlier in this series we have seen how to use Argo, Tekton and Knative to trigger a resource after getting input from Falcosidekick. 
+Recently, Falcosidekick got a new output type support for [Cloud Functions](https://github.com/falcosecurity/falcosidekick/pull/241). 
+
+In this part, we will learn how can use Falcosidekick + Cloud Functions to detect and delete a compromised pod.
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:23]: In this part, let us learn how we can use Falcosidekick and Cloud Functions to detect and delete a compromised pod.

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+We will not go through what Cloud Functions is in depth but, you can always find a good overview in the [official documentation](https://cloud.google.com/functions).
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:25]: We will not go through what Cloud Functions is in-depth, however, you can always find a good overview about it in the [official documentation](https://cloud.google.com/functions).

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+Here is the high-level overview architecture that shows what we want to achieve at the end of the day:
+
+![cloud_functions_reference_arch](/docs/images/cloud_functions_reference_arch.png)
+
+You can find all the related code and resources in [this repository](https://github.com/Dentrax/kubernetes-response-engine-based-on-gke-and-gcloudfunctions).
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:31]: The link is not working.

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+$ gcloud projects add-iam-policy-binding ${GOOGLE_PROJECT_ID} \
+--member="serviceAccount:${SA_ACCOUNT}@${GOOGLE_PROJECT_ID}.iam.gserviceaccount.com" \
+--role="roles/cloudfunctions.developer"
+
+$ gcloud projects add-iam-policy-binding ${GOOGLE_PROJECT_ID} \
+--member="serviceAccount:${SA_ACCOUNT}@${GOOGLE_PROJECT_ID}.iam.gserviceaccount.com" \
+--role="roles/cloudfunctions.invoker"
+```
+
+At the beginning, we already enabled [WorkloadIdentity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) feature for our GKE Cluster by setting `--workload-pool` flag. What we need to do here is we should add a `iam.workloadIdentityUser` role for the given Service Account.
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:67]: In the beginning, we already enabled [WorkloadIdentity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) feature for our GKE Cluster by setting `--workload-pool` flag. What we need to do here is to add a `iam.workloadIdentityUser` role for the given Service Account.

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+As the blog title said already, we need to create a [GKE cluster](https://cloud.google.com/kubernetes-engine) with workload identity enabled:
+
+```shell
+$ GOOGLE_PROJECT_ID=$(gcloud config get-value project)
+$ CLUSTER_NAME=falco-falcosidekick-demo
+$ gcloud container clusters create $CLUSTER_NAME --workload-pool ${GOOGLE_PROJECT_ID}.svc.id.goog
+$ gcloud container clusters get-credentials $CLUSTER_NAME
+```
+
+### Configuring IAM Service Accounts
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:51]: Configure IAM Service Accounts

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+```bash
+$ kubectl annotate serviceaccount \
+  --namespace $FALCO_NAMESPACE \
+  falco-falcosidekick \
+  iam.gke.io/gcp-service-account=${SA_ACCOUNT}@${GOOGLE_PROJECT_ID}.iam.gserviceaccount.com
+```
+
+### Create Necessary Cluster Role
+
+To limit function role access in the particular cluster, we need to ensure our SA only have limited permissions within a particular namespace by using [Role Bindings](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding). It should only access _Pod_ resource for _DELETE_ action:
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:86]: To limit function role access in the particular cluster, ensure that only SA has limited permissions within a particular namespace by using [Role Bindings](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding).

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+We can test it with [auth can-i](https://kubernetes.io/docs/reference/access-authn-authz/authorization/#checking-api-access) to check if roles are set correctly
+```bash
+$ kubectl auth can-i list deployments  # no
+$ kubectl auth can-i delete pod  # yes
+$ kubectl access-matrix  # github.com/corneliusweig/rakkess
+```
+
+### Create Secret Manager
+
+Where Secret Manager get involved our architecture is we had to find a way out to initialize our _kubeclient_ in our function. Simply, we need to store our `pod-destroyer`'s _KUBECONFIG_ and access from the function.
```
Comment (radhikapc) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:130]: It is unclear what you mean by "Where Secret Manager get involved our architecture". Please rewrite for clarity.

**Review by radhikapc — CHANGES_REQUESTED** [review]
Minor changes requested for clarity.

**Comment by Dentrax** [comment]
Thanks for the reviews @radhikapc 🧡 I pushed to the latest version.

**Review by radhikapc — APPROVED** [review]
LGTM

**Comment by developer-guy** [comment]
Hello folks @radhikapc @Dentrax, we have to wait for a release falcosidekick for its helm chart before publishing this blog post on the Falco website, because there is not support for GCP Cloud Functions in the current Falcosidekick Helm Chart yet 🙋🏻‍♂️

**Comment by Dentrax** [comment]
I think this PR is blocking from the following:

- In the beginning of the article, I [referenced](https://github.com/falcosecurity/falco-website/pull/453/files#diff-8a071d693ed114f0770acb26beb3389dcf5d0d41dd2f744528a9391e2c424a27R17) the 6th part (see the pull #446) of the KRE series.
- See, the https://github.com/falcosecurity/charts/pull/239 for the Cloud Function Helm Chart.

File: content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+
+> _This blog post is part of a series of articles about how to create a `Kubernetes` response engine with `Falco`, `Falcosidekick` and a `FaaS`._
+>
+> See other posts:
+>
+> * [Kubernetes Response Engine, Part 1 : Falcosidekick + Kubeless]({{< ref "/blog/falcosidekick-reponse-engine-part-1-kubeless" >}})
+> * [Kubernetes Response Engine, Part 2 : Falcosidekick + OpenFaas]({{< ref "/blog/falcosidekick-reponse-engine-part-2-openfass" >}})
+> * [Kubernetes Response Engine, Part 3 : Falcosidekick + Knative]({{< ref "/blog/falcosidekick-reponse-engine-part-3-knative" >}})
+> * [Kubernetes Response Engine, Part 4 : Falcosidekick + Tekton]({{< ref "/blog/falcosidekick-reponse-engine-part-4-tekton" >}})
+> * [Kubernetes Response Engine, Part 5 : Falcosidekick + Argo]({{< ref "/blog/falcosidekick-reponse-engine-part-5-argo" >}})
```
Comment (Issif) [inline content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:16]: Please add part 6 in that list, and part 7 in list of all other posts

Comment (Dentrax) [reply content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:16]: Thanks, done!

**Review by leogr — COMMENTED** [review]
The Netifly builder complained about the missing `/blog/falcosidekick-reponse-engine-part-6-cloud-run` reference:

```
10:41:36 AM: Building sites … ERROR 2021/06/29 08:41:36 [en] REF_NOT_FOUND: Ref "/blog/falcosidekick-reponse-engine-part-6-cloud-run": "/opt/build/repo/content/en/blog/falcosidekick-reponse-engine-part-7-cloud-functions.md:17:70": page not found
```

It's happening because the part 6 file is missing here. Rebasing this PR on top of the master branch should solve the issue.

**Comment by Dentrax** [comment]
Sure, rebased!

**Review by Issif — APPROVED** [review]

**Comment by developer-guy** [comment]
@radhikapc i think you should approve this PR too, can please approve the PR? 🙋🏻‍♂️🙏

**Review by radhikapc — APPROVED** [review]
Lgtm

**Pull request merged**